### ISSUE TITLE
copy and keep original folder structure

### ISSUE THREAD

**Issue by pajooh** [gulpjs/gulp#151]
i have a list of files to copy:

``` js
    gulp.src([list of js files])
      .pipe(gulp.dest(build_dir))
```

the files are copied, but flattened.
how can i reserve original directory structure? i can't find an option for that

**Comment by andreasonny83** [comment]
This will copy the file from inside the _temp folder into the dist folder keeping the directory structure:

gulp.task('copy_to_dist', function() {
  gulp.src([
    'index.html',
   '_temp/css/**/*',
    '_temp/img/**/*',
    '_temp/js/**/*'
  ])
    .pipe(gulp.dest(function(file) {
      var dest = file.base.replace('_temp', 'dist');
      return dest;
    }));
});

**Comment by DanielD** [comment]
I used this:

`gulp.task("copy-assets", function() {
    var assets = {
        js: [
            paths.npm + "es6-shim/es6-shim.min.js",
            paths.npm + "systemjs/dist/system-polyfills.js",
            paths.npm + "systemjs/dist/js/system.src.js",
            paths.npm + "bootstrap/dist/js/bootstrap.js",
            paths.npm + "angular2/es6/dev/src/testing/shims_for_IE.js",
            paths.npm + "angular2/bundles/angular2.dev.js",
            paths.npm + "angular2/bundles/router.dev.js",
            paths.npm + "angular2/bundles/angular2-polyfills.js",
            paths.npm + "angular2/bundles/http.dev.js",
            paths.npm + "rxjs/bundles/Rx.js",
            paths.npm + "typescript/lib/typescript.js"
        ],
        css: [
            paths.npm + "bootstrap/dist/css/bootstrap.css"
        ]
    };
    gulp.src(assets.js, { base: paths.npm })
        .pipe(gulp.dest(paths.npmOutput));
});`

where, paths.npm and paths.npmOutput are './node_modules/' and './wwwroot/node_modules/', respectively.

**Comment by slavafomin** [comment]
It's very counter-intuitive to me why Gulp flattens the filestructure when globs are not used. By the way I've tried to use `base` option of the `src` function, but it's not working. I've managed to make it work only after replacing `base` with `cwd` and using `**` in it. But for some reason, the official documentation doesn't mention `cwd` option. I'm using Gulp `3.9.1`,

**Comment by nekhaly** [comment]
Just to conclude in a simple statement :) .. Use `**/` in `src` to keep folder structure in `dest`
Like this
`gulp.src('**/views/*.html').pipe(gulp.dest('build'));`
considering `views` folder in the same level with `gulpfile.js`

**Comment by rlvandaveer** [comment]
If you are using an array of paths as the argument to src, the folder structure will not be respected using the pattern `**/`, e.g. `gulp.src(['media/*', 'media/norespect/**/']).pipe(gulp.dest('build'));` The destination will be flattened. It does work if it's the only pattern passed to src.

**Comment by oychao** [comment]
@jbdemonte Thank you so much.

**Comment by maltem-za** [comment]
@rlvandaveer I found this SO answer to be helpful: https://stackoverflow.com/a/32880264

**Comment by dennisbaskin** [comment]
### tl;dr

I needed to move one nested subdirectory and its contents out into my build directory, with a custom structure

### Working Solution

This is a fairly old thread but I found @andreasonny83 's answer to be the right path for what I was trying to accomplish.

What I needed is to have specific "component" directories that contained nested structures with demos moved to my `build/demos` directory. I have many components in a directory named `components`. Not all of the `component-...` subdirectories needed to be moved, only a select few. I needed to be able to copy the selected few, and avoid an additional `components` directory being added (when using base option). I was able to accomplish this through the following as suggested by @andreasonny83 :

```
gulp.task('task-name', () => {
  return gulp.src([
    'components/component1/**',
    'components/component2/**',
  ]).pipe(gulp.dest((file) => {
    return file.base.replace('components', 'build/demos')
  }))
})
``` 

The following shows the before (with components) and after (with the sibling build directory) file structure:

```
components/
├── component1/
│   └── subdir/
├── component2/
│   └── subdir/
├── component3-no-copy/
├── component4-no-copy/
build/
└── demos/
    ├── component1/
    │   └── subdir/
    └── component2/
        └── subdir/
```


If someone knows a more idiomatic approach to this, it would be great to know. But for now, this is decent workaround.

**Comment by jeffreyscottfrench** [comment]
## Simplified Version

This method is based off the response from @jbdemonte, and has worked for me, and also seems to be the most straightforward I've seen so far:

* define your input files/globs with paths relative to project base, either in a variable or inline in the gulp.src first parameter
* set your gulp.src base option to the source folder in the second parameter
* set your output folder as usual

Your code would look like this, where ./src is your source folder and ./dest is your destination folder:

```
var filesToCopy = ['./src/file.txt', './src/.hiddenFile', './src/subFolder/**/*.+(gif|png)'];

gulp.task('copyFiles'), function(){
    return gulp.src( filesToCopy, { base: './src' } )
    .pipe( gulp.dest( './dest' )
}
```

Hope that helps!
